### PR TITLE
feat(perf issues): Optimize issue search based on query

### PR DESCRIPTION
**Pull request by ceorourke**
If there's a specific query like `issue.category:performance` or `issue.type:performance_n_plus_one`, skip the query for error issues and vice versa for performance issues. If there is no specific query or the queries use both (like for error and performance issue types), we still query for both.

### CODE REVIEW

File: src/sentry/search/snuba/executors.py [hunk 1]
```diff
@@ -280,32 +283,42 @@ def snuba_search(
...
+            rows = snuba_error_results["data"]
+            total = snuba_error_results["totals"]["total"]
+            row_length = len(rows)
 
         organization = Organization.objects.get(id=organization_id)
-        if features.has("organizations:performance-issues", organization):
+        if (
+            features.has("organizations:performance-issues", organization)
+            and not group_categories
+            or GroupCategory.PERFORMANCE in group_categories
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:320]: We need parens here `(not group_categories or GroupCategory.PERFORMANCE in group_categories)`

File: src/sentry/search/snuba/executors.py [hunk 2]
```diff
@@ -280,32 +283,42 @@ def snuba_search(
...
-
-        rows = snuba_error_results["data"]
-        total = snuba_error_results["totals"]["total"]
-        row_length = len(rows)
+        group_categories = set()
+
+        for group_type in group_types:
+            for category, types in get_category_type_mapping().items():
+                if GroupType(group_type) in types:
+                    group_categories.add(category)
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:291]: Won't we also have `GroupCategory` values here from `issue.category`?

File: src/sentry/search/snuba/executors.py [hunk 3]
```diff
@@ -202,7 +201,10 @@ def snuba_search(
 
         conditions = []
         having = []
+        group_types = set()
         for search_filter in search_filters:
+            if search_filter.key.name in ("issue.category", "issue.type"):
+                group_types.update(value for value in search_filter.value.raw_value)
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:207]: Does it make sense to handle these separately?

```
if search_filter.key.name in ("issue.category"):
    group_categories.update(GroupCategory(value) for value in search_filter.value.raw_value)
if search_filter.key.name in ("issue.category", "issue.type"):
    group_categories.update(GROUP_TYPE_TO_CATEGORY[GroupType(value)] for value in search_filter.value.raw_value)
````

Comment (ceorourke) [reply src/sentry/search/snuba/executors.py:207]: Whether you query type or category, by the time it gets here it's always the type value

Comment (wedamija) [reply src/sentry/search/snuba/executors.py:207]: Ahh, I forgot about that. We could probably use the second if and save some complexity later on.

File: src/sentry/search/snuba/executors.py [hunk 4]
```diff
@@ -202,7 +201,13 @@ def snuba_search(
 
         conditions = []
         having = []
+        group_categories: Set[int] = set()
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:204]: This is a `Set[GroupCategory]`

**Review by wedamija — APPROVED** [review]

**Pull request merged**